### ISSUE TITLE
Facture number Current Year prefix

### ISSUE THREAD

**Issue by hgoncalves** [InvoiceShelf/InvoiceShelf#77]
Hi here in Switzerland were I live it's quite common to prefix the Documents (Invoices/Payments) with the current year like:
**INV-240001**

It would be really nice to have a check box in invoices settings next to the prefix input, to enable the current year to be used in addition with the text prefix

**Comment by azaricstefan** [comment]
Is there a requirement to have invoices start from 1 for each new year?

Like:
INV-240001 <----2024
INV-240002
INV-240003
INV-250001 <----2025
INV-250002
INV-250003
INV-260001 <----2026
INV-260002
INV-260003

**Comment by hgoncalves** [comment]
Yes, indeed this is a good point. It would be expected to start a new counter for each year. 
Btw let me know if there's something I can do to contribute to the project.

**Comment by azaricstefan** [comment]
Would you have time to craft a draft PR?
I can help with development and testing.

**Comment by hgoncalves** [comment]
Sure, I can give it a try over the weekend.

**Comment by kamelohr** [comment]
@hgoncalves Just installed invoiceshelf for the first time today and also was stumbling upon this issue. I'm running 1.2.2 via Docker and if I got that right, your use case should already be possible:

<img width="2035" alt="image" src="https://github.com/InvoiceShelf/InvoiceShelf/assets/6642871/eadb98b8-8cb7-4477-8829-061bc5c4dd00">

Just add a new `date format` component (lower right button) and move it to the wanted position. Then change the capital Y to a lower one. That should do the trick.

However, the "restart the counter every year at 1" function should be a new feature, if I didn't miss something. 

Additionally I'd like to ask if it would be possible to make the counter start at a specific number. Let's say I like to number my invoices continuously and after switching from another invoice software and I need to start at 42. Then it would be nice to be able to enter the starting number of the sequence. So, in addition to the yearly auto reset option, would a hard "reset now to 1 or my wanted number" option be really appreciated.

**Comment by enoversum** [comment]
> @hgoncalves Just installed invoiceshelf for the first time today and also was stumbling upon this issue. I'm running 1.2.2 via Docker and if I got that right, your use case should already be possible:
> 
> <img alt="image" width="2000" src="https://private-user-images.githubusercontent.com/6642871/334170192-eadb98b8-8cb7-4477-8829-061bc5c4dd00.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTczMzM4MDUsIm5iZiI6MTcxNzMzMzUwNSwicGF0aCI6Ii82NjQyODcxLzMzNDE3MDE5Mi1lYWRiOThiOC04Y2I3LTQ0NzctODgyOS0wNjFiYzVjNGRkMDAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MDJUMTMwNTA1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTZiY2E0YWY3MDhkZTExOTg3YTMyMzZhMmM0NGY2MWU2NzJmMDVmODM5ZjU3MmYwODQ4Y2VmZmM5ZTU3NTEzZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.ejnG1Zbsq-jMvNle6SKENRm52a6ea_JF3C6nxvndL3Y">
> Just add a new `date format` component (lower right button) and move it to the wanted position. Then change the capital Y to a lower one. That should do the trick.
> 
> However, the "restart the counter every year at 1" function should be a new feature, if I didn't miss something.
> 
> Additionally I'd like to ask if it would be possible to make the counter start at a specific number. Let's say I like to number my invoices continuously and after switching from another invoice software and I need to start at 42. Then it would be nice to be able to enter the starting number of the sequence. So, in addition to the yearly auto reset option, would a hard "reset now to 1 or my wanted number" option be really appreciated.

This was already a missing feature with Crater. You can manually edit `app/Services/SerialNumberFormatter.php` and add the following line after line 129 to make InvoiceShelf re-start counting invoices/estimates etc. on a new year:

`->whereYear('created_at', date('Y'))`.

The whole block looks like this then: 

public function setNextSequenceNumber()
    {
        $companyId = $this->company;

        $last = $this->model::orderBy('sequence_number', 'desc')
            ->where('company_id', $companyId)
            ->where('sequence_number', '<>', null)
            ->whereYear('created_at', date('Y'))
            ->take(1)
            ->first();

        $this->nextSequenceNumber = ($last) ? $last->sequence_number + 1 : 1;

        return $this;
    }

**Comment by hgoncalves** [comment]
@enoversum thanks for your input. I didn't manage to have a look at it yet but I would like to make it optional and be able to have this as a setting.

**Comment by enoversum** [comment]
> @enoversum thanks for your input. I didn't manage to have a look at it yet but I would like to make it optional and be able to have this as a setting.

Absolutely, that's just a dirty fix from Crater times, very much a hack, as this file would get overwritten on a future update.

**Comment by rihards-simanovics** [comment]
I won't close the issue but tentatively relate it to #252.
